go/nomdl/parser.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2017 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package nomdl
+
+import (
+	"io"
+	"strconv"
+	"strings"
+	"text/scanner"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/types"
+)
+
+// Parser provides ways to parse Noms types.
+type Parser struct {
+	lex *lexer
+}
+
+// ParserOptions allows passing options into New.
+type ParserOptions struct {
+	// Filename is the name of the file we are currently parsing.
+	Filename string
+}
+
+// New creates a new Parser.
+func New(r io.Reader, options ParserOptions) *Parser {
+	s := scanner.Scanner{}
+	s.Filename = options.Filename
+	s.Mode = scanner.ScanIdents | scanner.ScanComments | scanner.SkipComments
+	s.Init(r)
+	lex := lexer{scanner: &s}
+	return &Parser{&lex}
+}
+
+// ParseType parses a string describing a Noms type.
+func ParseType(code string) (typ *types.Type, err error) {
+	p := New(strings.NewReader(code), ParserOptions{})
+	err = catchSyntaxError(func() {
+		typ = p.parseType()
+		p.ensureAtEnd()
+	})
+	return
+}
+
+// MustParseType parses a string describing a Noms type and panics if there
+// is an error.
+func MustParseType(code string) *types.Type {
+	typ, err := ParseType(code)
+	d.PanicIfError(err)
+	return typ
+}
+
+func (p *Parser) ensureAtEnd() {
+	p.lex.eat(scanner.EOF)
+}
+
+// Type :
+//   TypeWithoutUnion (`|` TypeWithoutUnion)*
+//
+// TypeWithoutUnion :
+//   `Blob`
+//   `Bool`
+//   `Number`
+//   `String`
+//   `Type`
+//   `Value`
+//   CycleType
+//   ListType
+//   MapType
+//   RefType
+//   SetType
+//   StructType
+//
+// CycleType :
+//   `Cycle` `<` uint32 `>`
+//
+// ListType :
+//   `List` `<` Type? `>`
+//
+// MapType :
+//   `Map` `<` (Type `,` Type)? `>`
+//
+// RefType :
+//   `Set` `<` Type `>`
+//
+// SetType :
+//   `Set` `<` Type? `>`
+//
+// StructType :
+//   `struct` StructName? `{` StructFields? `}`
+//
+// StructFields :
+//   StructField
+//   StructField `,` (StructField `,`)*
```

Comment:
You are right. This still didn't end up correct.

Your suggestion should work.